### ISSUE TITLE
[Feature] Store DeepSeek-V4.1 C1/C2 Main KV in packed FP4 on Hopper

### ISSUE THREAD

**Issue by zhangxiaolei123456** [sgl-project/sglang#38902]
### Checklist

- [ ] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [ ] Please use English. Otherwise, it will be closed.

### Motivation

baseline: #38798 

DeepSeek-V4.1 already applies E2M1 fake quantization to C1/C2 compressed latents, but the result is dequantized and requantized into the legacy FlashMLA FP8 cache:

```text
E2M1 fake quant
  -> dequantized BF16/FP32 values
  -> FP8 E4M3 quantization
  -> legacy FlashMLA page (584 logical bytes per compressed token)
```

This preserves FP4 numerical behavior but not FP4 storage density. This issue proposes a versioned packed layout for **DeepSeek-V4.1 C1/C2 Main KV only**. SWA KV, Indexer KV, V4 C4/C128, and request-scoped C2 pair state remain unchanged.

The implementation is split into a correctness-first staged consumer and a later native SM90 mixed-cache consumer. The first two phases provide the persistent KV-capacity gain without requiring an immediate FlashMLA kernel rewrite.

## Scope

In scope:

- Main KV written by V4.1 ratio-1 and ratio-2 `kv_source_layer_ids`
- C1/C2 decode and prefill stores
- Memory-pool sizing and typed layout descriptors
- Dense/sparse prefill and decode consumers
- HiCache and Mooncake/NIXL PD page transfer
- Hopper/H20 software E2M1 encode/decode
- A future SM100 hardware-conversion specialization

Out of scope:

- SWA KV layout
- FP4 Indexer KV layout
- V4 C4/C128 layout
- C2 pending-pair ring layout or transfer semantics
- Radix-tree logical slot/page indexing

## Current Behavior

The V4.1 C1/C2 fused kernels perform:

1. C1 projection, or C2 pair pooling.
2. RMSNorm over 512 channels.
3. RoPE over the final 64 channels.
4. E2M1 fake quantization with one E4M3 scale per 16 channels.
5. Requantization of the 448 noPE channels to FP8 E4M3 with one UE8M0 scale per 64 channels.
6. Storage in the legacy FlashMLA page layout.

The legacy logical row is:

```text
448 B noPE FP8 + 128 B RoPE BF16 + 8 B FP8 scale area = 584 B
```

The pool additionally pads each physical page to a multiple of 576 bytes. At the default page size 256, the allocated size is 149,760 bytes, or 585 bytes per slot.

## Proposed Layout ABI

Use a page-level structure-of-arrays layout backed by one contiguous
`torch.uint8` allocation:

```text
DSV41_MAIN_KV_E2M1_BLOCK16_ROPE_BF16_V1
├── payload : [page_slots, 224] uint8
│             448 E2M1 values, even channel in low nibble
├── scales  : [page_slots,  32] uint8
│             first 28 bytes are positive E4M3 scales
│             last 4 bytes are reserved and must be zero
└── rope    : [page_slots,  64] bfloat16
              128 bytes
```

Recommended physical size: **384 bytes per compressed slot**.

The 32-byte scale stride is intentional for SM90:

- Every scale row is 32-byte aligned.
- Payload rows are 224 bytes and RoPE rows are 128 bytes, both naturally vectorizable.
- For a 256-slot C1 page, all region boundaries and the 98,304-byte page are 4 KiB aligned.
- For a 128-slot C2 page, all region boundaries and the 49,152-byte page are also 4 KiB aligned.

Default page-size examples:

| Pool | page slots | payload offset | scale offset | RoPE offset | page bytes |
|---|---:|---:|---:|---:|---:|
| C1 | 256 | 0 | 57,344 | 65,536 | 98,304 |
| C2 | 128 | 0 | 28,672 | 32,768 | 49,152 |

A compact 380-byte variant with `[page_slots, 28]` scales would save another 4 bytes per slot, but each scale row would have a 28-byte stride. For H20 the aligned 384-byte ABI is preferred unless benchmarks show that the compact layout wins end to end.

Do not expose `torch.float4_e2m1fn_x2` in the runtime ABI. The payload and scale regions remain `uint8` so serialization, host storage, and RDMA registration do not depend on framework FP4 dtype support.

### Exact numerical contract

For each 16-channel block:

```text
amax  = max(abs(x))
scale = E4M3_RNE(clamp(amax / 6, 2^-9, 448))
code  = E2M1_RNE_SATFINITE(x / scale)
value = decode_e2m1(code) * float(scale)
```

Requirements:

- E2M1 conversion is round-to-nearest-even and saturates to `[-6, 6]`.
- Adding `+0.0f` before conversion clears negative zero, matching the current fake-quant path.
- Scale bytes are the raw positive E4M3 encoding, not UE8M0 exponents.
- Nibble order is part of the ABI: channel `2i` uses bits `[3:0]`, channel `2i+1` uses bits `[7:4]`.
- RoPE channels keep current semantics: apply block-16 E2M1 fake quantization, dequantize, then store the resulting 64 values as BF16. No RoPE scale is persisted.
- Reserved scale bytes are zero to keep serialization deterministic.

SM90 must use an integer/LUT E2M1 encoder and decoder and must not depend on Blackwell FP4 conversion instructions. SM100 can add a separately dispatched `cvt.rn.satfinite.e2m1x2` specialization with bitwise-equivalent output.

## Runtime Types and Memory Pool

Add an explicit layout descriptor instead of inferring semantics from tensor dtype or shape:

```python
class MainKVLayout(enum.Enum):
    FLASHMLA_FP8_V1 = "flashmla_fp8_v1"
    DSV41_E2M1_BLOCK16_ROPE_BF16_V1 = (
        "dsv41_e2m1_block16_rope_bf16_v1"
    )


@dataclass(frozen=True)
class MainKVLayoutSpec:
    layout: MainKVLayout
    page_slots: int
    payload_bytes_per_slot: int
    scale_bytes_per_slot: int
    rope_bytes_per_slot: int
    bytes_per_slot: int
    payload_offset: int
    scale_offset: int
    rope_offset: int
    page_bytes: int


@dataclass(frozen=True)
class PackedMainKVView:
    storage: torch.Tensor
    spec: MainKVLayoutSpec
```

Add `DeepSeekV41PackedMainKVPool(KVCache)` with:

```python
def get_view(self, local_layer_id: int) -> PackedMainKVView: ...
def get_page_buffer(self, local_layer_id: int) -> torch.Tensor: ...
def get_payload(self, local_layer_id: int) -> torch.Tensor: ...
def get_scales(self, local_layer_id: int) -> torch.Tensor: ...
def get_rope(self, local_layer_id: int) -> torch.Tensor: ...
def get_bytes_per_page(self) -> int: ...
def get_bytes_per_slot(self) -> int: ...
```

The backing tensor is `[num_pages, spec.page_bytes]` contiguous `uint8`. Region accessors return views over that allocation and must not allocate.

`DeepSeekV4LayerItem` should carry an explicit `main_kv_layout`. Add a typed
`get_extra_key_view(layer_id)` API for new code. Keep
`get_extra_key_buffer(layer_id)` only as the legacy FlashMLA-FP8 accessor so a
packed buffer can never be silently interpreted as a 584-byte row.

In `_init_compressed_pools()`, select the packed pool only when all of these are
true:

- model type is `deepseek_v41`;
- compression ratio is 1 or 2;
- the selected Main-KV layout is packed FP4;
- the backend has a staged or direct consumer.

All V4 C4/C128 pools and all SWA pools continue to use `DeepSeekV4SingleKVPool`.

## Memory Accounting

The estimator should query pool/layout descriptors:

```python
bytes_per_slot = pool.layout_spec.bytes_per_slot
bytes_per_full_token = sum(
    bytes_per_slot / compress_ratio
    for each PP-local V4.1 kv_source layer
)
```

Remove the assumption that every compressed latent costs 584 bytes from `DSV4PoolConfigurator`. Use physical `page_bytes / page_slots` when estimating capacity so allocator padding is included.

For a checkpoint with one C1 source and three C2 sources:

```text
legacy logical bytes/raw token = 584 * (1 + 3/2) = 1,460
packed aligned bytes/raw token = 384 * (1 + 3/2) =   960
saving                         =                       500 B
```

This saves approximately **476.8 MiB per million raw tokens per GPU** for Main KV alone. The 380-byte compact variant would save 510 B/token, or 486.4 MiB per million tokens. Indexer KV, SWA KV, allocator metadata, and request state are not included.

## Store Kernels

Add layout-specific entry points:

```text
flash_c1_decode_pack_main_kv_fp4
flash_c2_decode_pack_main_kv_fp4
flash_c1_prefill_pack_main_kv_fp4
flash_c2_prefill_pack_main_kv_fp4
```

The existing C1/C2 projection, pair pooling, RMSNorm, BF16 rounding, RoPE, and PDL ordering remain unchanged. Only replace the final fake-quant-plus-FP8-store epilogue.

### C1

- One CTA per token, 256 threads for the 512-channel latent.
- Preserve the current pre-RoPE BF16 latent publication for Indexer `wk`.
- Compute RMSNorm across all 512 channels.
- Apply RoPE to channels `[448:512]`.
- Compute one E4M3 scale and 16 E2M1 codes per noPE block.
- Store 224 payload bytes, 28 scale bytes, four zero padding bytes, and 128 BF16 RoPE bytes into their page regions.
- `out_loc <= 0` publishes no cache data, preserving CUDA-graph padding.

### C2

- Preserve the request-scoped pair-state ring and its current speculative overwrite semantics.
- Even position: write only the pending pair state.
- Odd position: combine `(pos-1, pos)`, RMSNorm, publish the pre-RoPE latent, use RoPE position `pos-1`, and store at `raw_out_loc // 2`.
- Chunk-boundary completion, rejected draft replay, and padding behavior must remain unchanged.

The prefill wrappers may share the same pack epilogue and layout descriptor, but should retain separate launch entry points where metadata shape or graph capture differs.

## Phase 1 Consumer: Staged Compatibility

Implement correctness-first adapters for both decode and prefill:

```text
packed Main KV
  -> gather only selected C1/C2 slots
  -> decode E2M1 and load BF16 RoPE
  -> reproduce the legacy 584-byte FP8 representation in a reusable workspace
  -> call the existing FlashMLA consumer
```

The adapter must reproduce the current post-fake-quant FP8 cache contract:

- Decode noPE E2M1 values with their block-16 E4M3 scales.
- Recompute the legacy per-64 UE8M0 scale.
- Quantize the decoded noPE values to FP8 E4M3.
- Copy the BF16 RoPE tail unchanged.
- Write the legacy 8-byte scale area deterministically.

This allows direct output comparison against the current implementation and avoids unpacking the full persistent cache.

### Decode workspace and index remapping

The existing sparse FlashMLA API expects a paged `extra_k_cache` plus global slot indices. Therefore the adapter should:

1. Allocate one reusable, CUDA-graph-stable workspace sized for the captured
   maximum `batch * aligned_topk`.
2. Assign each query a disjoint contiguous workspace range.
3. Gather valid source slots from `extra_indices_in_kvcache`.
4. Convert those rows into legacy FlashMLA pages.
5. Rewrite valid indices to workspace slots and preserve `-1` padding.
6. Pass the remapped indices and existing `extra_topk_length` to FlashMLA.

The workspace is transient and shared across layers. It must not be attached to request slots or persisted in Radix/HiCache.

For source-sharing layers, cache the staged conversion only when the source layout, selected-index tensor generation, and top-k lengths are identical. Correctness must not rely on this reuse; it is an optional optimization.

### Prefill

Extend the existing sparse-prefill gather/dequant workspace:

- BF16 sparse prefill decodes packed rows directly to the existing BF16 workspace.
- Q8KV8 sparse prefill decodes and requantizes directly to the existing FP8 workspace.
- Dense/small-prefill paths use the same remapped legacy workspace as decode until a direct consumer exists.

## Phase 2 Consumer: Native SM90 Mixed Cache

Add a versioned API instead of overloading `is_fp8_kvcache`:

```cpp
flash_mla_with_mixed_kvcache(
    q,
    swa_fp8_cache,
    swa_indices,
    main_cache_bytes,
    main_indices,
    main_layout,
    main_page_slots,
    ...
)
```

The SM90 extra-cache load pipeline should:

1. Resolve `(page, slot)` from each sparse Main-KV index.
2. Cooperatively load the 224-byte payload, 32-byte scale row, and 128-byte BF16 RoPE row.
3. Decode low/high nibbles through a constant/register E2M1 LUT.
4. Convert each raw E4M3 scale byte to FP32 and multiply its 16 channels.
5. Materialize BF16/FP8 fragments in registers or shared memory.
6. Feed the existing QK and V computation.
7. Merge SWA and Main-KV contributions with the existing online-softmax state.

FP4 is a storage format on H20; the attention math remains in the existing supported compute type. Native FP4 Tensor Core support is not required.

Acceptance for the direct path should require a lower TPOT than the staged path and no regression against legacy FP8 at short contexts where Main-KV bandwidth is not dominant.

## HiCache

V4.1 low-ratio Main-KV entries are already transferred in FULL-page units and marked `page_aligned_only`. Keep that invariant.

The assembler should consume `pool.get_page_buffer()` and `pool.get_bytes_per_page()` rather than relying on the legacy FlashMLA row
shape. `DeepSeekV4PagedHostPool` is already parameterized by `item_bytes`; it does not need to understand E2M1. Token-granular
`transfer_cache_dsv4_mla()` must never be used for this SoA layout.

Required tests:

- D2H/H2D whole-page round trip for C1 and C2.
- `layer_first`, `page_first`, and supported direct layouts.
- Radix prefix hit, eviction, reload, and slot reuse.
- Storage-backend serialization round trip.
- Rejection of non-page-aligned transfer indices.

## PD Disaggregation

Logical FULL-page IDs and compressed slot mapping remain unchanged:

```text
C1 slot = full_slot
C2 slot = full_slot // 2
```

Mooncake/NIXL should register and transfer the complete contiguous packed page. Existing `get_contiguous_buf_infos()` can continue to expose one pointer and one item length per source buffer, but the bootstrap contract needs an explicit layout signature. `item_len` alone is not a sufficient semantic identifier.

Suggested signature:

```python
KVRegionLayout(
    kind="dsv41_main_kv",
    layout_id="dsv41_e2m1_block16_rope_bf16_v1",
    compression_ratio=1 | 2,
    global_page_size=256,
    page_slots=256 | 128,
    page_bytes=98304 | 49152,
    source_layer_id=...,
)
```

Prefill and decode must validate the ordered region signatures before registration or transfer. A mismatch is a hard error. Never infer or reinterpret a packed page from dtype, tensor shape, or `item_len`.

This signature should travel through the common registration model so both Mooncake and NIXL enforce the same rule. C2 pending-row transfer remains a separate request-scoped state component and is not part of this Main-KV page.

## Configuration and Fallback

Use separate layout and consumer choices:

```text
--dsv41-main-kv-layout auto|flashmla_fp8|packed_fp4
--dsv41-main-kv-consumer auto|staged|direct
```

Recommended rollout:

- Initial default: `layout=flashmla_fp8`.
- Explicit validation: `layout=packed_fp4, consumer=staged`.
- After SM90 qualification: `layout=auto, consumer=auto` selects packed/direct on supported Hopper builds.
- SM100 may select a hardware conversion store specialization while retaining the same page ABI.
- Unsupported backends fall back to `flashmla_fp8` only during local configuration resolution, with one visible log message.
- A received packed PD/HiCache buffer must never silently fall back or be interpreted as FP8.
- P and D endpoints must resolve the same layout; consumer choice may differ because it does not change persistent bytes.

### Unit and kernel correctness

- Exhaustive E2M1 code/decode table and nibble ordering.
- RNE midpoint cases, saturation, subnormal E4M3 scales, scale floor, NaN/Inf policy, positive/negative zero.
- C1 pack/unpack against the existing fake-quant reference.
- C2 even/odd positions, request interleaving, chunk boundaries, `out_loc=0`, ring wrap, and rejected speculative tokens.
- Reserved scale bytes remain zero.
- SM90 software conversion and SM100 specialization produce identical bytes.

### Attention correctness

- Legacy FP8 versus packed-staged Main KV at the byte-reconstructed legacy workspace boundary.
- Attention output max/mean error and logits KL divergence.
- C1 and each C2 source independently.
- Dense decode, sparse decode, dense prefill, BF16 sparse prefill, and Q8KV8 sparse prefill.
- CUDA eager, decode graph replay, prefill graph replay, DP attention, CP, PP, and source-sharing layers.

### Cache and distributed correctness

- Radix prefix hit/miss, partial page, eviction, and slot reuse.
- HiCache D2H/H2D and storage round trip.
- Mooncake and NIXL same-layout transfer.
- Negative tests for page size, layout ID, page bytes, region order, and source-layer mismatch.
- PD prefill without speculative decode to DSpark decode with multiple gamma values, including C2 pending-row handoff.
- Cancellation during transfer and destination-slot reuse.

### H20 performance

Benchmark legacy FP8, packed-staged, and packed-direct with:

- batch sizes 1, 8, 32, and 64;
- context lengths 4K, 32K, and 128K;
- C1 and C2 source layers separately;
- CUDA graph on/off;
- TTFT, TPOT, store-kernel time, unpack time, attention time, HBM bytes,
  workspace peak, and achieved bandwidth.

Also compare the 380-byte compact scale stride against the recommended 384-byte aligned stride before freezing the v1 ABI.

### PR 1: Layout, pool, store, and reference conversion

- Layout enum/spec and typed views
- `DeepSeekV41PackedMainKVPool`
- C1/C2 packed store kernels for SM90
- Pack/unpack references and kernel tests
- Layout-driven memory estimator
- Feature flags, with legacy FP8 still default

### PR 2: Staged consumer, HiCache, and PD

- Decode gather/unpack/requant adapter and index remapping
- Dense and sparse prefill adapters
- CUDA-graph-stable reusable workspaces
- HiCache page integration
- Common PD layout signature and Mooncake/NIXL validation
- End-to-end correctness and failure-mode tests

This PR makes packed FP4 usable and delivers the persistent capacity benefit.

### PR 3: Native SM90 mixed SWA-FP8/Main-FP4 attention

- Direct packed Main-KV load pipeline
- Online-softmax merge with unchanged SWA FP8 input
- Hopper tuning and staged/direct benchmarks
- Auto-selection on qualified H20/SM90 configuratio

**Comment by minseoc03** [comment]
Hi, are you already planning to work on PR 1? If not, I'd be interested in picking up part of it, especially the packed layout/store path.

**Comment by kjoshiDTX** [comment]
Hi @zhangxiaolei123456! This would be my first open source contribution and I would be interested in working on PR #2 if anyone isn't already.

**Comment by yuyu5333** [comment]
Hi, i would like to support pr3's work. Has it already started?
